mllib/src/main/scala/org/apache/spark/ml/feature/PolynomialMapper.scala

```diff
@@ -0,0 +1,137 @@
+}
+
+object PolynomialMapper {
+  /**
+   * The number that combines k items from N items without repeat, i.e. the binomial coefficient.
+   */
+  private def binomialCoefficient(N: Int, k: Int): Int = {
+    (N - k + 1 to N).product / (1 to k).product
+  }
+
+  /**
+   * The number of monomials of a `numVariables` vector after expanding at a specific polynomial
+   * degree `degree`.
+   */
+  private def numMonomials(degree: Int, numVariables: Int): Int = {
+    binomialCoefficient(numVariables + degree - 1, degree)
+  }
+
+  /**
+   * The number of monomials of a `numVariables` vector after expanding from polynomial degree 1 to
+   * polynomial degree `degree`.
+   */
+  private def numExpandedDims(degree: Int, numVariables: Int): Int = {
+    binomialCoefficient(numVariables + degree, numVariables) - 1
+  }
+
+  /**
+   * Multiply two polynomials, the first is the original vector, i.e. the expanded vector with
+   * degree 1, while the second is the expanded vector with degree `currDegree - 1`. A new expanded
+   * vector with degree `currDegree` will be generated after the function call.
+   *
+   * @param lhs original vector with degree 1
+   * @param rhs expanded vector with degree `currDegree - 1`
+   * @param nDim the dimension of original vector
+   * @param currDegree the polynomial degree that need to be achieved
+   */
+  private def expandVector(lhs: Vector, rhs: Vector, nDim: Int, currDegree: Int): Vector = {
+    (lhs, rhs) match {
+      case (l: DenseVector, r: DenseVector) =>
+        val rLen = rhs.size
+        val allExpansions = l.toArray.zipWithIndex.flatMap { case (lVal, lIdx) =>
+          val numToKeep = numMonomials(currDegree - 1, nDim - lIdx)
+          r.toArray.slice(rLen - numToKeep, rLen).map(rVal => lVal * rVal)
+        }
+        Vectors.dense(allExpansions)
+
+      case (SparseVector(lLen, lIdx, lVal), SparseVector(rLen, rIdx, rVal)) =>
+        val len = numMonomials(currDegree, nDim)
+        var numToKeepCum = 0
+        val allExpansions = lVal.zip(lIdx).flatMap { case (lv, li) =>
+          val numToKeep = numMonomials(currDegree - 1, nDim - li)
+          val currExpansions = rVal.zip(rIdx).map { case (rv, ri) =>
+            val realIdx = ri - (rLen - numToKeep)
+            (if (realIdx >= 0) lv * rv else 0.0, numToKeepCum + realIdx)
+          }
+          numToKeepCum += numToKeep
+          currExpansions
+        }.filter(_._1 != 0.0)
+        Vectors.sparse(len, allExpansions.map(_._2), allExpansions.map(_._1))
+
+      case _ => throw new Exception("vector types are not match.")
+    }
+  }
+
+  /**
+   * Transform a vector of variables into a larger vector which stores the polynomial expansion from
+   * degree 1 to degree `degree`.
+   */
+  private def transform(degree: Int)(feature: Vector): Vector = {
+    val nDim = feature.size
+    feature match {
+      case f: DenseVector =>
+        (2 to degree).foldLeft(Array(feature.copy)) { (vectors, currDegree) =>
```

Comment:
The implementation is too expensive. We already know the output size and it should be easy to fill in the values one by one.